Macro should not require Clone by default

This code:

```
//! src/main.rs
use multi_index_map::MultiIndexMap;

#[derive(MultiIndexMap, Debug)]
struct Order {
    #[multi_index(hashed_unique)]
    order_id: u32,
    #[multi_index(ordered_unique)]
    timestamp: u64,
    #[multi_index(hashed_non_unique)]
    trader_name: String,
}

fn main() {}
```

produces this error:

```
error[E0599]: no method named `clone` found for mutable reference `&mut Order` in the current scope
 --> src\main.rs:3:10
  |
3 | #[derive(MultiIndexMap, Debug)]
  |          ^^^^^^^^^^^^^ method not found in `&mut Order`
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the trait `Clone` defines an item `clone`, but is explicitly unimplemented
  = note: this error originates in the derive macro `MultiIndexMap` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Order: Clone` is not satisfied
 --> src\main.rs:3:10
  |
3 | #[derive(MultiIndexMap, Debug)]
  |          ^^^^^^^^^^^^^ the trait `Clone` is not implemented for `Order`
  |
  = note: required because of the requirements on the impl of `Clone` for `Slab<Order>`
  = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider annotating `Order` with `#[derive(Clone)]`
  |
4 | #[derive(Clone)]
  |

Some errors have detailed explanations: E0277, E0599.
For more information about an error, try `rustc --explain E0277`.
```

This is unintuitive and undesireable because I'm not cloning any `Order`s nor is the functionality of the multi-index map tied to cloning `Order`s. Its understandable for *each key* type to be cloneable because of the nature of multiple index maps directing to the same store. However, it seems the macro imposes `Clone` on the generated map data structure for reasons unknown to me.

It should probably implement `Clone` only if `Order` implements `Clone`.

--- a/README.md
+++ b/README.md
@@ -32,14 +32,15 @@ Current implementation supports:
 
 This crate provides a derive macro `MultiIndexMap`, which when applied to the struct representing an element will generate a map to store and access these elements.
 Annotations are used to specify which fields to index. Currently `hashed_unique`, `hashed_non_unique`, `ordered_unique`, and `ordered_non_unique` are supported.
-The element must implement `Clone`.
+The types of all indexed fields must implement `Clone`.
+If the MultiIndexMap needs to be cloned, `Clone` must be implemented manually, see `examples/main.rs` for an example of how to do this.
 
 ## Example
 
 ```rust
 use multi_index_map::MultiIndexMap;
 
-#[derive(MultiIndexMap, Clone, Debug)]
+#[derive(MultiIndexMap, Debug)]
 struct Order {
     #[multi_index(hashed_unique)]
     order_id: u32,

--- a/RELEASES.md
+++ b/RELEASES.md
@@ -1,3 +1,9 @@
+Version 0.8.0 (2023-08-30)
+==========================
+
+- Remove `Clone` requirement on elements, now only the indexed fields must implement Clone. This should be helpful when storing non-Clonable types in un-indexed fields.
+- If the MultiIndexMap does need to be Cloned, this must be implemented manually, however this should be fairly simple to do next to where the element is defined. See `examples/main.rs`.
+
 Version 0.7.1 (2023-08-30)
 ==========================
 

--- a/multi_index_map/Cargo.toml
+++ b/multi_index_map/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "multi_index_map"
-version = "0.7.1"
+version = "0.8.0"
 edition = "2021"
 authors = ["Louis Wyborn <louiswyborn@gmail.com>"]
 rust-version = "1.62"
@@ -14,7 +14,7 @@ readme = "README.md"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-multi_index_map_derive = { version = "0.7.1", path = "../multi_index_map_derive" }
+multi_index_map_derive = { version = "0.8.0", path = "../multi_index_map_derive" }
 
 # Used as the backing store of all the elements.
 slab = { version = "0.4" }

--- a/multi_index_map/benches/performance.rs
+++ b/multi_index_map/benches/performance.rs
@@ -16,6 +16,18 @@ pub struct TestElementWithOnlyIndexedFields {
     field_ordered_non_unique: u32,
 }
 
+impl Clone for MultiIndexTestElementWithOnlyIndexedFieldsMap {
+    fn clone(&self) -> Self {
+        Self {
+            _store: self._store.clone(),
+            _field_hashed_unique_index: self._field_hashed_unique_index.clone(),
+            _field_hashed_non_unique_index: self._field_hashed_non_unique_index.clone(),
+            _field_ordered_unique_index: self._field_ordered_unique_index.clone(),
+            _field_ordered_non_unique_index: self._field_ordered_non_unique_index.clone(),
+        }
+    }
+}
+
 const BENCH_SIZES: &[u32] = &[100u32, 1000u32, 10000u32, 100000u32];
 
 fn insert_benchmark(c: &mut Criterion) {

--- a/multi_index_map/examples/main.rs
+++ b/multi_index_map/examples/main.rs
@@ -13,6 +13,18 @@ mod inner {
         pub(crate) trader_name: String,
         pub(crate) note: String,
     }
+
+    // Manually implement Clone, this can be auto generated correctly by rust-analyzer
+    impl Clone for MultiIndexOrderMap {
+        fn clone(&self) -> Self {
+            Self {
+                _store: self._store.clone(),
+                _order_id_index: self._order_id_index.clone(),
+                _timestamp_index: self._timestamp_index.clone(),
+                _trader_name_index: self._trader_name_index.clone(),
+            }
+        }
+    }
 }
 
 fn main() {

--- a/multi_index_map/tests/capacity_manipulations.rs
+++ b/multi_index_map/tests/capacity_manipulations.rs
@@ -1,6 +1,6 @@
 use multi_index_map::MultiIndexMap;
 
-#[derive(MultiIndexMap, Clone, Debug)]
+#[derive(MultiIndexMap, Debug)]
 struct TestElement {
     #[multi_index(hashed_unique)]
     field1: i32,

--- a/multi_index_map/tests/get_and_modify_mut.rs
+++ b/multi_index_map/tests/get_and_modify_mut.rs
@@ -3,7 +3,7 @@ use multi_index_map::MultiIndexMap;
 #[derive(Hash, PartialEq, Eq, Clone)]
 struct TestNonPrimitiveType(u64);
 
-#[derive(MultiIndexMap, Clone)]
+#[derive(MultiIndexMap)]
 struct TestElement {
     #[multi_index(hashed_non_unique)]
     field1: usize,

--- a/multi_index_map/tests/hashed_non_unique.rs
+++ b/multi_index_map/tests/hashed_non_unique.rs
@@ -3,7 +3,7 @@ use multi_index_map::MultiIndexMap;
 #[derive(Hash, PartialEq, Eq, Clone, Debug)]
 struct TestNonPrimitiveType(u64);
 
-#[derive(MultiIndexMap, Clone, Debug)]
+#[derive(MultiIndexMap, Debug)]
 struct TestElement {
     #[multi_index(hashed_non_unique)]
     field1: TestNonPrimitiveType,

--- a/multi_index_map/tests/hashed_unique.rs
+++ b/multi_index_map/tests/hashed_unique.rs
@@ -3,7 +3,7 @@ use multi_index_map::MultiIndexMap;
 #[derive(Hash, PartialEq, Eq, Clone)]
 struct TestNonPrimitiveType(u64);
 
-#[derive(MultiIndexMap, Clone)]
+#[derive(MultiIndexMap)]
 struct TestElement {
     #[multi_index(hashed_unique)]
     field1: TestNonPrimitiveType,

--- a/multi_index_map/tests/iter_after_modify.rs
+++ b/multi_index_map/tests/iter_after_modify.rs
@@ -1,6 +1,6 @@
 use multi_index_map::MultiIndexMap;
 
-#[derive(MultiIndexMap, Debug, Clone)]
+#[derive(MultiIndexMap, Debug)]
 pub(crate) struct Order {
     #[multi_index(hashed_unique)]
     pub(crate) order_id: u32,

--- a/multi_index_map/tests/mixed_non_unique.rs
+++ b/multi_index_map/tests/mixed_non_unique.rs
@@ -1,6 +1,6 @@
 use multi_index_map::MultiIndexMap;
 
-#[derive(MultiIndexMap, Clone)]
+#[derive(MultiIndexMap)]
 struct MultipleOrderedNonUniqueStruct {
     #[multi_index(ordered_non_unique)]
     field1: u32,

--- a/multi_index_map/tests/ordered_non_unique.rs
+++ b/multi_index_map/tests/ordered_non_unique.rs
@@ -3,7 +3,7 @@ use multi_index_map::MultiIndexMap;
 #[derive(Hash, PartialEq, Eq, Clone, Debug, PartialOrd, Ord)]
 struct TestNonPrimitiveType(u64);
 
-#[derive(MultiIndexMap, Clone, Debug)]
+#[derive(MultiIndexMap, Debug)]
 struct TestElement {
     #[multi_index(ordered_non_unique)]
     field1: TestNonPrimitiveType,

--- a/multi_index_map/tests/ordered_unique.rs
+++ b/multi_index_map/tests/ordered_unique.rs
@@ -3,7 +3,7 @@ use multi_index_map::MultiIndexMap;
 #[derive(Hash, PartialEq, Eq, Clone, PartialOrd, Ord)]
 struct TestNonPrimitiveType(u64);
 
-#[derive(MultiIndexMap, Clone)]
+#[derive(MultiIndexMap)]
 struct TestElement {
     #[multi_index(ordered_unique)]
     field1: TestNonPrimitiveType,

--- a/multi_index_map/tests/reverse_iter.rs
+++ b/multi_index_map/tests/reverse_iter.rs
@@ -1,5 +1,5 @@
 use multi_index_map::MultiIndexMap;
-#[derive(MultiIndexMap, Clone, PartialEq, Debug)]
+#[derive(MultiIndexMap, PartialEq, Debug)]
 struct TestElement {
     #[multi_index(ordered_non_unique)]
     field1: usize,

--- a/multi_index_map/tests/update.rs
+++ b/multi_index_map/tests/update.rs
@@ -3,7 +3,7 @@ use multi_index_map::MultiIndexMap;
 #[derive(Hash, PartialEq, Eq, Clone)]
 struct TestNonPrimitiveType(u64);
 
-#[derive(MultiIndexMap, Clone, PartialEq, Debug)]
+#[derive(MultiIndexMap, PartialEq, Debug)]
 struct TestElement {
     #[multi_index(hashed_non_unique)]
     field1: i32,

--- a/multi_index_map_derive/Cargo.toml
+++ b/multi_index_map_derive/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "multi_index_map_derive"
-version = "0.7.1"
+version = "0.8.0"
 edition = "2021"
 authors = ["Louis Wyborn <louiswyborn@gmail.com>"]
 rust-version = "1.62"

--- a/multi_index_map_derive/src/generators.rs
+++ b/multi_index_map_derive/src/generators.rs
@@ -813,7 +813,7 @@ pub(crate) fn generate_expanded(
     lookup_table_fields_reserve: impl Iterator<Item = proc_macro2::TokenStream>,
 ) -> proc_macro2::TokenStream {
     quote! {
-        #[derive(Default, Clone)]
+        #[derive(Default)]
         #element_vis struct #map_name {
             _store: ::multi_index_map::slab::Slab<#element_name>,
             #(#lookup_table_fields)*

